po/vi.po

```diff
@@ -6188,57 +6198,75 @@ msgid ""
 "to [grapheme clusters](https://docs.rs/unicode-segmentation/latest/"
 "unicode_segmentation/struct.Graphemes.html)."
 msgstr ""
+"`String::chars` trả về một con trỏ chạy dọc theo các ký tự của chuỗi. Lưu ý "
+"rằng một `char` có thể khác với một \"ký tự\" thông thường bởi vì một khái "
+"niệm gọi là [grapheme clusters](https://docs.rs/unicode-segmentation/latest/"
+"unicode_segmentation/struct.Graphemes.html)."
 
 #: src/std-types/string.md
 msgid ""
 "When people refer to strings they could either be talking about `&str` or "
 "`String`."
-msgstr ""
+msgstr "Khi nói đến chuỗi, ta có thể sử dụng `&str` hoặc `String` đều được."
 
 #: src/std-types/string.md
 msgid ""
 "When a type implements `Deref<Target = T>`, the compiler will let you "
 "transparently call methods from `T`."
 msgstr ""
+"Khi một kiểu dữ liệu implement `Deref<Target = T>`, trình biên dịch sẽ cho "
+"phép gọi các hàm thuộc về `T` trên kiểu dữ liệu đó."
 
 #: src/std-types/string.md
 msgid ""
 "We haven't discussed the `Deref` trait yet, so at this point this mostly "
 "explains the structure of the sidebar in the documentation."
 msgstr ""
+"Chúng ta chưa đi sâu vào `Deref` trait, việc đề cập đến `Deref` ở đây chủ "
+"yếu là để giải thích cấu trúc của mục lục tài liệu."
 
 #: src/std-types/string.md
 msgid ""
 "`String` implements `Deref<Target = str>` which transparently gives it "
 "access to `str`'s methods."
 msgstr ""
+"`String` implement `Deref<Target = str>`, nên các hàm của `str` đều có thể "
+"được gọi trên các biến kiểu `String`."
 
 #: src/std-types/string.md
 msgid "Write and compare `let s3 = s1.deref();` and `let s3 = &*s1;`."
 msgstr ""
+"So sánh `let s3 = s1.deref();` và `let s3 = &*s1;`. Cái nào dễ hiểu hơn?"
 
 #: src/std-types/string.md
 msgid ""
 "`String` is implemented as a wrapper around a vector of bytes, many of the "
 "operations you see supported on vectors are also supported on `String`, but "
 "with some extra guarantees."
 msgstr ""
+"`String` được được viết dựa trên một vector gồm nhiều byte, nên những hàm "
+"gọi được trên vector cũng có thể được gọi trên `String`, nhưng đi kèm với "
+"một số điều kiện để đảm bảo các tính chất của chuỗi."
 
 #: src/std-types/string.md
 msgid "Compare the different ways to index a `String`:"
-msgstr ""
+msgstr "So sánh các cách truy cập phần tử trong một biến kiểu `String`:"
 
 #: src/std-types/string.md
 msgid ""
 "To a character by using `s3.chars().nth(i).unwrap()` where `i` is in-bound, "
 "out-of-bounds."
 msgstr ""
+"Truy cập một ký tự bằng cách sử dụng `s3.chars().nth(i).unwrap()` với `i` "
+"nằm trong phạm vi của chuỗi, và với `i` nằm ngoài phạm vi của chuỗi."
 
 #: src/std-types/string.md
 msgid ""
 "To a substring by using `s3[0..4]`, where that slice is on character "
 "boundaries or not."
 msgstr ""
+"Truy cập một chuỗi con bằng cách sử dụng `s3[0..4]`, với phạm vi của chuỗi "
```

Comment:
"xâu con" might be better than "chuỗi con" (just a suggestion). Instead of using "nằm trên" try out "nằm ngoài"? Cái chuỗi này nằm ngoài chuỗi kia sounds a bit more logical ?

Reply:
@qu-ngx Ah, I can see where the confusion coming from. Rust string has this rather strange property that, unlike other language like C++, when you access a string with `s[i]` syntax, the returned value could be an invalid character. The reason is, Rust's string is made of unicode characters, so each character could span multiple index. Suppose you have the string `s = "Xin chào"`. All the `X`, `i`, `n`, ` `, `c`, `h`, `o` are ASCII characters, therefore it need only one byte, consuming one slot in the string. However, `à` is a unicode character, so it need 2 bytes, consuming 2 slot. What this means is that when you try to access "Xin chào", Rust allow you to access `s[0..2]`, which prints out `Xi`, `s[5..8]`, which equal to `hà`, but you can't access `s[5..7]`, since 7 is in the middle of `à` character. This is what the excerpt refer to as the `character boundaries`.
However, I acknowledge that my translation poorly conveys that idea. I will see if I can improve this.